src/navier_stokes/VCINSStaggeredHierarchyIntegrator.cpp

```diff
@@ -0,0 +1,3295 @@
+    // with a coupled advection-diffusion solver.
+    if (d_adv_diff_hier_integrator)
+    {
+        VariableDatabase<NDIM>* var_db = VariableDatabase<NDIM>::getDatabase();
+        const int U_adv_diff_current_idx =
+            var_db->mapVariableAndContextToIndex(d_U_adv_diff_var, d_adv_diff_hier_integrator->getCurrentContext());
+        if (isAllocatedPatchData(U_adv_diff_current_idx))
+        {
+            copy_side_to_face(U_adv_diff_current_idx, d_U_current_idx, d_hierarchy);
+        }
+    }
+    return;
+} // initializePatchHierarhcy
+
+void
+VCINSStaggeredHierarchyIntegrator::preprocessIntegrateHierarchy(const double current_time,
+                                                                const double new_time,
+                                                                const int num_cycles)
+{
+    INSHierarchyIntegrator::preprocessIntegrateHierarchy(current_time, new_time, num_cycles);
+
+    const int coarsest_ln = 0;
+    const int finest_ln = d_hierarchy->getFinestLevelNumber();
+    const double dt = new_time - current_time;
+    const double half_time = current_time + 0.5 * dt;
+
+    // Keep track of the number of cycles to be used for the present integration
+    // step.
+    if (!d_creeping_flow && (d_current_num_cycles == 1) &&
+        (d_convective_time_stepping_type == MIDPOINT_RULE || d_convective_time_stepping_type == TRAPEZOIDAL_RULE))
+    {
+        TBOX_ERROR(d_object_name << "::preprocessIntegrateHierarchy():\n"
+                                 << "  time stepping type: "
+                                 << enum_to_string<TimeSteppingType>(d_convective_time_stepping_type)
+                                 << " requires num_cycles > 1.\n"
+                                 << "  at current time step, num_cycles = "
+                                 << d_current_num_cycles
+                                 << "\n");
+    }
+
+    // Allocate the scratch and new data.
+    for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
+    {
+        Pointer<PatchLevel<NDIM> > level = d_hierarchy->getPatchLevel(ln);
+        level->allocatePatchData(d_scratch_data, current_time);
+        level->allocatePatchData(d_new_data, new_time);
+        level->allocatePatchData(d_velocity_C_idx, current_time);
+        level->allocatePatchData(d_velocity_rhs_C_idx, current_time);
+        level->allocatePatchData(d_velocity_D_idx, current_time);
+        level->allocatePatchData(d_velocity_rhs_D_idx, current_time);
+        level->allocatePatchData(d_pressure_D_idx, current_time);
+        level->allocatePatchData(d_pressure_rhs_D_idx, current_time);
+        level->allocatePatchData(d_temp_sc_idx, current_time);
+        level->allocatePatchData(d_temp_cc_idx, current_time);
+        level->allocatePatchData(d_rho_interp_idx, current_time);
+        level->allocatePatchData(d_mu_interp_idx, current_time);
+    }
+
+    // Compute variable mass density.
+    if (d_rho_var)
+    {
+        if (d_rho_fcn)
```

Comment:
rho should not be initialized by registered CartGridFunction here. rho is evolved. rho should be set only at time t = 0, and this is done automatically by the base class HierarchyIntegrator::initializeLevelData(). As written this will always be called, because a user will register CartGridFunction to set rho(x,y,z,t=0) almost 100%.